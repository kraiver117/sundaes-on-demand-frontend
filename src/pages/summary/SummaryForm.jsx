import React, { useState } from 'react';
import { Form, Button } from 'react-bootstrap';

export const SummaryForm = () => {
  const [checked, setChecked] = useState(false);

  const checkboxLabel = (
    <span>
      I agree to <span style={{ color: 'blue' }}>Terms and Conditions</span>
    </span>
  );

  return (
    <Form>
      <Form.Group controlId='terms-and-conditions'>
        <Form.Check
          type='checkbox'
          checked={checked}
          onChange={(e) => setChecked(e.target.checked)}
          label={checkboxLabel}
        />
      </Form.Group>
      <Button variant='primary' type='submit' disabled={!checked}>
        Confirm Order
      </Button>
    </Form>
  );
};
