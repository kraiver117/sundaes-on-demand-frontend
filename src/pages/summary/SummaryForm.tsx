import React, { FormEvent, FC, useState } from "react";
import { Form, Button, Popover, OverlayTrigger } from "react-bootstrap";
import { SetOrderPhase } from "../../interfaces/phase";

export const SummaryForm: FC<SetOrderPhase> = ({
  setOrderPhase,
}: SetOrderPhase) => {
  const [checked, setChecked] = useState(false);

  function handleSubmit(e: FormEvent<HTMLFormElement>) {
    e.preventDefault();

    // pass along to the next phase
    setOrderPhase("completed");
  }

  const popover = (
    <Popover id="popover-basic">
      <Popover.Body>No ice cream will actually be delivered</Popover.Body>
    </Popover>
  );

  const checkboxLabel = (
    <span>
      I agree to
      <OverlayTrigger placement="right" overlay={popover}>
        <span style={{ color: "blue" }}>Terms and Conditions</span>
      </OverlayTrigger>
    </span>
  );

  return (
    <Form onSubmit={handleSubmit}>
      <Form.Group controlId="terms-and-conditions">
        <Form.Check
          type="checkbox"
          checked={checked}
          onChange={(e) => setChecked(e.target.checked)}
          label={checkboxLabel}
        />
      </Form.Group>
      <Button variant="primary" type="submit" disabled={!checked}>
        Confirm Order
      </Button>
    </Form>
  );
};
