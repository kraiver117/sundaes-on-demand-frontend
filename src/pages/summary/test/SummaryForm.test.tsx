import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { SummaryForm } from "../SummaryForm";
import userEvent from "@testing-library/user-event";

test("should render Summary Form correctly", () => {
  const { asFragment } = render(<SummaryForm setOrderPhase={() => {}} />);
  expect(asFragment()).toMatchSnapshot();
});

test("Initial State", () => {
  render(<SummaryForm setOrderPhase={() => {}} />);

  const checkBox = screen.getByRole("checkbox", {
    name: /terms and conditions/i,
  });
  expect(checkBox).not.toBeChecked();

  const confirmButton = screen.getByRole("button", { name: /confirm order/i });
  expect(confirmButton).toBeDisabled();
});

test("Checkbox enables button on first click and disables on second click", () => {
  render(<SummaryForm setOrderPhase={() => {}} />);
  const checkBox = screen.getByRole("checkbox", {
    name: /terms and conditions/i,
  });
  const confirmButton = screen.getByRole("button", { name: /confirm order/i });

  userEvent.click(checkBox);
  expect(confirmButton).toBeEnabled();

  userEvent.click(checkBox);
  expect(confirmButton).toBeDisabled();
});

test("popover respond to hover", async () => {
  render(<SummaryForm setOrderPhase={() => {}} />);

  // popover starts out hidden
  const nullPopover = screen.queryByText(
    /no ice cream will actually be delivered/i
  );
  expect(nullPopover).not.toBeInTheDocument();

  // popover appears upon mouseover of checkbox label
  const termsAndConditions = screen.getByText(/terms and conditions/i);
  userEvent.hover(termsAndConditions);

  const popover = screen.getByText(/no ice cream will actually be delivered/i);
  expect(popover).toBeInTheDocument();

  // popover disappears when we mouse out
  userEvent.unhover(termsAndConditions);
  await waitForElementToBeRemoved(() =>
    screen.queryByText(/no ice cream will actually be delivered/i)
  );
});
