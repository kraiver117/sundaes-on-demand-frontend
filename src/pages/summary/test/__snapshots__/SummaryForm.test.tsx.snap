// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Summary Form correctly 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="terms-and-conditions"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="terms-and-conditions"
          title=""
        >
          <span>
            I agree to
            <span
              style="color: blue;"
            >
              Terms and Conditions
            </span>
          </span>
        </label>
      </div>
    </div>
    <button
      class="btn btn-primary"
      disabled=""
      type="submit"
    >
      Confirm Order
    </button>
  </form>
</DocumentFragment>
`;
